### ISSUE TITLE
mboot.efi fails with static IP

### ISSUE THREAD

**Issue by oizone** [vmware/esx-boot#6]
I'm running an environment which doesn't have DHCP available and need to use static IP for the UEFI HTTP boot.
Dell R640 servers boot fine with static IP configured, but downloading boot.cfg fails as it seems that mboot.efi tries to initiate DHCP.
Boot fails with error:
Error in Dhcp4->Start: Timeout
Error getting IPv4 address: Timeout
Configuration error while parsing http://...file
Fatal error: 15 (Not found)

Couldn't find any documentation on how to use mboot.efi with static IP HTTP UEFI boot, is this possible?
If it's not possible now, I'd like to make it a feature request.

**Comment by TimothyPMann** [comment]
That should work, though it may not have been tested much. If has_ipv4_addr() returns EFI_SUCCESS the first time (and every time) it is called, DHCPv4 is not run. So if Dell provided a UI in the setup dialogs or the UEFI shell that lets you assign a static IP address, and you create a boot option that has an explicit boot URL in it, I would expect it to work. (If you only assign a static IP in ESXi itself, that won't be visible to the bootloader.)

I don't have a Dell machine handy, but when time permits I can try this on an HPE machine and confirm whether if it works there as intended.

**Comment by oizone** [comment]
Yes, I've configured following redfish properties in BIOS through ansible:
        PxeDev1EnDis: "Disabled"
        PxeDev2EnDis: "Disabled"
        PxeDev3EnDis: "Disabled"
        PxeDev4EnDis: "Disabled"
        HttpDev1EnDis: "Enabled"
        HttpDev2EnDis: "Disabled"
        HttpDev3EnDis: "Disabled"
        HttpDev4EnDis: "Disabled"
        HttpDev1VlanEnDis: "Enabled"
        HttpDev1DnsDhcpEnDis: "Disabled"
        HttpDev1DhcpEnDis: "Disabled"
        HttpDev1Uri: "http://10.37.0.26/{{ fqdn }}/mboot.efi"
        HttpDev1Protocol: "IPv4"
        HttpDev1VlanId: "{{ vlan }}"
        HttpDev1Dns1: "{{ dns1 }}"
        HttpDev1Dns2: "{{ dns2 }}"
        HttpDev1Ip: "{{ ip }}"
        HttpDev1Mask: "{{ netmask }}"
        HttpDev1Gateway: "{{ gateway }}"

So specific properties are that it's static ip in tagged VLAN. As said loading mboot.efi works completely fine, so static IP itself works. Once mboot.efi starts to load boot.cfg, it fails saying dhcp timed out. I can give access to Dell server in our lab if that helps.

**Comment by TimothyPMann** [comment]
Hmm, that sounds familiar. Actually the only reason that mboot.efi has code that can initiate DHCP at all is that I hit an issue (even with dynamic IP) where, although the firmware's UEFI boot manager had acquired an IP address and used it to load mboot.efi, once mboot.efi started, the NIC didn't have an IP address that mboot could ue. It sounds like the same thing happened to you when you configured a static IP: somehow the static IP is available only to the UEFI boot manager, not to a bootloader that the boot manager starts.

**Comment by oizone** [comment]
So, is this an issue in the firmware or in mboot.efi?
Meaning that does Dell need to fix this? Does this work on HPE for example?

**Comment by TimothyPMann** [comment]
I don't know if it works on HPE; I'll have to find time to try it. Deconfiguring the IP address before starting a bootloader might be a generic limitation or even an intentional feature of UEFI, though it is not documented in the UEFI spec. To understand why the IP address isn't there and find out if there is a way to recover it, I probably will also need to read the generic open source upstream UEFI code that the hardware vendors base their UEFI implementations on (namely Tianocore/EDK2). That is time consuming for me as there is a lot of code there that I'm not deeply familiar with, and it's not written in a very readable style. I am currently working on multiple other projects unrelated to UEFI HTTP boot, so I probably can't get to any of this soon.

**Comment by oizone** [comment]
If this is a by design limitation of UEFI HTTP boot, find it strange but let's assume it for a minute, then I assume it would be possible to pass the information through the uri? Bootloader already works different depending on it's name, bootx64.efi vs mboot.efi, so I assume similar model can pass the expected static ip info to it?
using http://server/mboot.efi?ip=10.10.10.10&mask=255.255.255.0 as uri?
Or if it can't be paramters, then workaround with the name, like http://server/mboot.10.10.10.10-255.255.255.0.efi
wildcard on http server side can fix and always still return same original mboot.efi

**Comment by oizone** [comment]
And as info, I've opened severity 2 support request on VMware support on this, id 21206887703.
I'll open a Dell support request also, if needed.

**Comment by TimothyPMann** [comment]
The bootloader doesn't work differently depending on its name. The bootloader doesn't have access to the URL it was loaded as.

**Comment by oizone** [comment]
Hmm, thought I had different behaviour in some of my tests when I didn't rename the bootloader to mboot.efi but used the original name bootx64.efi. But must have but something else, validated now and behaviour is the same with both names. Quite many changes ongoing at once in my lab.
I'm opening a support ticket on this to Dell also, to validate that their firmware is working as per UEFI HTTP design on this part.

**Comment by TimothyPMann** [comment]
It occurs to me that there's another option that may work for you.  ESXi also supports network booting directly from the installer ISO image, if your UEFI firmware is new enough (anything that supports UEFI HTTP boot should work). We probably should have documented this explicitly, since it's simpler to set up. Just put the ISO image on an HTTP server and configure the target machine to UEFI HTTP boot from the image's URL. When you do this, the firmware loads the entire image into memory and mboot accesses it as a ramdisk; mboot doesn't need an IP address.

Of course, when ESXi comes up, it won't have an IP address either. I don't know how you were planning to deal with that.

**Comment by TimothyPMann** [comment]
BTW, I looked at your SR 21206887703, and it seems that at some point VMware support could no longer reach you because your emails were bouncing due apparently to a forwarding loop. So if you stopped hearing updates at some point, that's why.

**Comment by oizone** [comment]
Using the ISO method would of course fix the boot itself, but since the point is in automated installation where each host requires it's own kickstart file, it would get very complicated since each host would need their own dedicated ISO. Doable sure, but in my opinion a bit messy workaround. Or do you see any other way of injecting the host specific kickstart file in this process?
Regarding the SR, I'm not working for TietoEVRY as an internal employee anymore, so my email in that SR is not valid anymore. You can update my contact email there as onni@rautanen.me , if that SR is still alive and it helps to push this forward through that one.

**Comment by vopop123** [comment]
Hi guys, did you find any solution for this issue? I'm facing the same one. Would be great, if you could share your findings. Thanks

**Comment by TimothyPMann** [comment]
Unfortunately I haven't found time to work on it. It would need a significant code change to mboot.

________________________________

**Comment by TimothyPMann** [comment]
I may have at last found a fix for this. I'm still testing, and it has definitely missed ESXi 8.0.3, but if all goes well it will be available eventually.

**Comment by TimothyPMann** [comment]
A fix is checked in and should become available in ESXi 9.0

**Comment by TimothyPMann** [comment]
ESXi 9.0 isn't out yet, though it's coming soon. The fix turned out to be fairly simple. Instead of asking UEFI to use the default local IPv4 address, which somehow is not set even though the UEFI boot manager just acquired an address and used it to load mboot itself, we now parse the device path that mboot was loaded from (which we were doing anyway for other reasons) find the IPv4 address in it, and ask UEFI to use that local IPv4 address. Voila, it works. So I was able to completely delete our code that triggers a new DHCP transaction.

As to the timeout you mention, that seems completely unrelated and I don't know why it would happen. However I gather you are working through another support channel about that and there's no need to discuss it here in this github bug report.

**Comment by TimothyPMann** [comment]
Also, having iPXE in the mix makes things very different. If iPXE is present, mboot will use iPXE's HTTP/TCP/IP stack instead of UEFI's native stack, so the original issue in this bug report shouldn't even be possible, though there could be any number of iPXE issues -- iPXE is very complex.

**Comment by omniproc** [comment]
I was investigating the mentioned behavior and suspected something along the lines. @mcb30 commented on it here: https://github.com/ipxe/ipxe/discussions/1369

Seems like iPXE choose to do the exact opposite (not use the device path) for reasons mentioned in the linked discussion. While I agree it could be very well iPXE related i'd like to point out: it works when using VMware EFI (nested VM) just fine, which is at least interesting.
